### PR TITLE
159 twinning protocol implementation

### PR DESCRIPTION
**Pull request by vedina**
implements #159 

- TwinningComponent in src/ramanchada2/protocols/twinning.py
- added test spectra for TiO2 used for the twinning
- introduces new class SpectraFrame (a dataframe with filenames and metadata , optionally spectra loaded in Spectrum)
- Metadata is essential for the twinning. 

- skeleton classes for metadata readers from different sources

### CODE REVIEW

**Comment by vedina** [comment]
I did not intend cherrypicking, this branch was rebased recently

**Comment by kerberizer** [comment]
rebased

**Comment by vedina** [comment]
> rebased

thanks

**Comment by kerberizer** [comment]
I'll take a look at those flake8 problems, but will likely skip the mypy errors unless the resolution is obvious to me.

File: src/ramanchada2/protocols/calibration.py [hunk 1]
```diff
@@ -191,7 +191,12 @@ def _plot(self, ax, **kwargs):
             basefmt=" ",
             label="Reference {}".format(self.sample),
         )
-        ax.set_xlabel("{}".format(self.ref_units))
+
+        if self.ref_units == "cm-1":
+            _units = rf"$\mathrm{{[{_units}]}}$"
```
Comment (kerberizer) [inline src/ramanchada2/protocols/calibration.py:196]: `_units` in the right side of the assignment seems undefined at this point.

File: src/ramanchada2/protocols/twinning.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+        spe: Spectrum,
+        boundaries=None,
+        prominence_coeff=0.01,
+        no_fit=False,
+        peak_intensity="height",
+    ):
+        try:
+            if boundaries is None:
+                boundaries = (self.reference_band_nm - 50, self.reference_band_nm + 50)
+            spe = spe.trim_axes(method="x-axis", boundaries=boundaries)
```
Comment (kerberizer) [inline src/ramanchada2/protocols/twinning.py:102]: mypy: error: "Spectrum" has no attribute "trim_axes"  [attr-defined]

File: src/ramanchada2/protocols/twinning.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+            prominence = spe.y_noise_MAD() * prominence_coeff
+            candidates = spe.find_peak_multipeak(prominence=prominence)
```
Comment (kerberizer) [inline src/ramanchada2/protocols/twinning.py:104]: error: "Spectrum" has no attribute "find_peak_multipeak"  [attr-defined]

File: src/ramanchada2/protocols/twinning.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+            fit_res = spe.fit_peak_multimodel(
```
Comment (kerberizer) [inline src/ramanchada2/protocols/twinning.py:105]: error: "Spectrum" has no attribute "fit_peak_multimodel"  [attr-defined]

File: src/ramanchada2/protocols/twinning.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    def process(
+        self, spe: SpectraFrame, source="spectrum", target="spectrum_harmonized"
+    ):
+        spe.multiply(self.correction_factor, source=source, target=target)
+        spe.spe_area(
+            source=target, target="area_harmonized", boundaries=self.boundaries
+        )
+
+    def process_spectrum(self, spe: Spectrum):
+        return spe * self.correction_factor
```
Comment (kerberizer) [inline src/ramanchada2/protocols/twinning.py:170]: error: Unsupported operand types for * ("Spectrum" and "float")  [operator]

**Review by kerberizer — CHANGES_REQUESTED** [review]
I've fixed the trivial flake8 problems, but the one remaining and the mypy problems still need fixing.

File: src/ramanchada2/protocols/calibration.py (hunk 1 above)
Comment (vedina) [reply src/ramanchada2/protocols/calibration.py:196]: should be self.ref_units at the right

File: src/ramanchada2/protocols/twinning.py (hunk 2 above)
Comment (vedina) [reply src/ramanchada2/protocols/twinning.py:102]: this is decorator issue

File: src/ramanchada2/protocols/twinning.py (hunk 3 above)
Comment (vedina) [reply src/ramanchada2/protocols/twinning.py:104]: decorator issue

File: src/ramanchada2/protocols/twinning.py (hunk 4 above)
Comment (vedina) [reply src/ramanchada2/protocols/twinning.py:105]: decorator issue

File: src/ramanchada2/protocols/twinning.py (hunk 5 above)
Comment (vedina) [reply src/ramanchada2/protocols/twinning.py:170]: remove process_spectrum

**Review by kerberizer — APPROVED** [review]
LGTM, with two remarks:
- it would be great to fix those MyPy problems in a more elegant way instead of instructing MyPy to simply ignore them;
- it would be great to have 100% test coverage, but it's understandable that it isn't always possible or practical.

**Comment by georgievgeorgi** [comment]
We agreed that MyPy problems will be fixed on later time with a major refactoring. For the time being we keep the  # ignore. Are there something else to be done or shall we proceed with merging this branch?

**Comment by vedina** [comment]
From my side I would say let's merge

**Review by georgievgeorgi — APPROVED** [review]

**Pull request merged**